### PR TITLE
Fix some bugs in image conversion

### PR DESCRIPTION
**Pull request by loheagn**
This patch fixes some bugs in image conversion using `LayerConvertFunc` and `ConvertHookFunc`.

1. append `nydus.remoteimage.v1` to `os.features` of the manifest descriptor in the index
2. add `"containerd.io/snapshot/nydus-blob": "true"` to the annotations of the blob layer descriptor in the manifest
3. avoid appending duplicate bootstrap layer digest to the diffID list of the config

This patch also adds a smoke test `TestContainerdImageConvert` to make sure `LayerConvertFunc` and `ConvertHookFunc` work well with the containerd image conversion library.

Signed-off-by: Nan Li <loheagn@icloud.com>

### CODE REVIEW

File: pkg/converter/convert_unix.go [hunk 1]
```diff
@@ -592,89 +592,130 @@ func LayerConvertFunc(opt PackOption) converter.ConvertFunc {
...
-			return nil, errors.Wrap(err, "merge nydus layers")
+// convertIndex modifies the original index by appending "nydus.remoteimage.v1"
+// to the Platform.OSFeatures of each modified manifest descriptors.
+func convertIndex(ctx context.Context, cs content.Store, orgDesc ocispec.Descriptor, newDesc *ocispec.Descriptor) (*ocispec.Descriptor, error) {
+	var orgIndex ocispec.Index
+	if _, err := readJSON(ctx, cs, &orgIndex, orgDesc); err != nil {
+		return nil, errors.Wrap(err, "read target image index json")
+	}
+	// isManifestModified is a function to check whether the manifest is modified.
+	isManifestModified := func(manifest ocispec.Descriptor) bool {
```
Comment (imeoer) [inline pkg/converter/convert_unix.go:617]: Maybe we can check whether it is nydus manifest (name it `isNydusManifest`) by recognizing bootstrap annotation, which would be safer?

File: pkg/converter/convert_unix.go [hunk 2]
```diff
@@ -592,89 +592,130 @@ func LayerConvertFunc(opt PackOption) converter.ConvertFunc {
...
+	config.RootFS.DiffIDs = []digest.Digest{}
+	for i, layer := range manifest.Layers {
+		config.RootFS.DiffIDs = append(config.RootFS.DiffIDs, digest.Digest(layer.Annotations[LayerAnnotationUncompressed]))
+		// Remove useless annotation.
+		delete(layer.Annotations, LayerAnnotationUncompressed)
+		manifest.Layers[i] = layer
+	}
+	if opt.Backend != nil {
+		config.RootFS.DiffIDs = []digest.Digest{digest.Digest(bootstrapDesc.Annotations[LayerAnnotationUncompressed])}
+	}
```
Comment (imeoer) [inline pkg/converter/convert_unix.go:702]: Be better:

```
if opt.Backend != nil {
	config.RootFS.DiffIDs = []digest.Digest{digest.Digest(bootstrapDesc.Annotations[LayerAnnotationUncompressed])}
} else {
	config.RootFS.DiffIDs = []digest.Digest{}
	for i, layer := range manifest.Layers {
		config.RootFS.DiffIDs = append(config.RootFS.DiffIDs, digest.Digest(layer.Annotations[LayerAnnotationUncompressed]))
		// Remove useless annotation.
		delete(layer.Annotations, LayerAnnotationUncompressed)
		manifest.Layers[i] = layer
	}
}
```

**Review by imeoer — COMMENTED** [review]
Thanks for the contribution!

File: pkg/converter/convert_unix.go (hunk 1 above)
Comment (loheagn) [reply pkg/converter/convert_unix.go:617]: Then we need to read json from the content store for each manifest descriptor.

Comment (loheagn) [reply pkg/converter/convert_unix.go:617]: Maybe we can store the modified manifest digest in a map in the manifest conversion hook?

File: pkg/converter/convert_unix.go (hunk 2 above)
Comment (loheagn) [reply pkg/converter/convert_unix.go:703]: Fixed.

File: pkg/converter/convert_unix.go (hunk 1 above)
Comment (imeoer) [reply pkg/converter/convert_unix.go:617]: Well, just checking if the manifest is changed should also work.

File: pkg/converter/convert_unix.go [hunk 3]
```diff
@@ -592,89 +592,131 @@ func LayerConvertFunc(opt PackOption) converter.ConvertFunc {
...
+		return nil, errors.Wrap(err, "read image config")
+	}
+	if opt.Backend != nil {
+		config.RootFS.DiffIDs = []digest.Digest{digest.Digest(bootstrapDesc.Annotations[LayerAnnotationUncompressed])}
+	} else {
+		config.RootFS.DiffIDs = make([]digest.Digest, 0, len(manifest.Layers))
+		for i, layer := range manifest.Layers {
 			config.RootFS.DiffIDs = append(config.RootFS.DiffIDs, digest.Digest(layer.Annotations[LayerAnnotationUncompressed]))
 			// Remove useless annotation.
 			delete(layer.Annotations, LayerAnnotationUncompressed)
```
Comment (imeoer) [inline pkg/converter/convert_unix.go:700]: Just use `delete(manifest.Layers[i].Annotations, LayerAnnotationUncompressed)`.

Comment (loheagn) [reply pkg/converter/convert_unix.go:700]: OK

**Review by imeoer — APPROVED** [review]

**Comment by loheagn** [comment]
@imeoer Can you please rerun the action? There may be a network error. Thanks :)

**Comment by changweige** [comment]
The current main branch has Golang race problem, which fails this PR. I will fix it

**Comment by changweige** [comment]
Please rebase to the latest main branch.

**Review by imeoer — APPROVED** [review]

**Pull request merged**